nodes/core/logic/18-sort.js

```diff
@@ -0,0 +1,182 @@
+                if (xp > yp) return dir;
+                return -dir;
+            };
+        };
+
+        function send_group(group) {
+            var key = key_is_payload
+                ? function(msg) { return msg.payload; }
+                : function(msg) {
+                    return eval_jsonata(node, key_exp, msg);
+                };
+            var comp = gen_comp(key);
+            var msgs = group.msgs;
+            try {
+                msgs.sort(comp);
+            }
+            catch (e) {
+                return; // not send when error
+            }
+            for (var i = 0; i < msgs.length; i++) {
+                var msg = msgs[i];
+                msg.parts.index = i;
+                node.send(msg);
+            }
+        };
+
+        function sort_payload(msg) {
+            var payload = msg.payload;
+            if (Array.isArray(payload)) {
+                var key = key_is_payload
+                    ? function(elem) { return elem; }
+                    : function(elem) {
+                        return eval_jsonata(node, key_exp, elem);
+                    };
+                var comp = gen_comp(key);
+                try {
+                    payload.sort(comp);
+                }
+                catch (e) {
+                    return false;
+                }
+                return true;
+            }
+            return false;
+        }
+
+        function check_parts(parts) {
+            if (parts.hasOwnProperty("id") &&
+                parts.hasOwnProperty("index")) {
+                return true;
+            }
+            return false;
+        }
+        
+        function process_msg(msg) {
+            if (!msg.hasOwnProperty("parts")) {
+                if (sort_payload(msg)) {
+                    node.send(msg);
+                }
+                return;
+            }
+            var parts = msg.parts;
+            if (!check_parts(parts)) {
+                return;
+            }
+            var gid = parts.id;
+            if (!pending.hasOwnProperty(gid)) {
+                pending[gid] = {
+                    count: undefined,
+                    msgs: []
+                };
+            }
+            var group = pending[gid];
+            var msgs = group.msgs;
+            msgs.push(msg);
+            if (parts.hasOwnProperty("count")) {
+                group.count = parts.count;
+            }
+            pending_count++;
+            if (group.count === msgs.length) {
+                delete pending[gid]
+                send_group(group);
+                pending_count -= msgs.length;
+            }
+            var max_msgs = max_kept_msgs_count(node);
+            if ((max_msgs > 0) && (pending_count > max_msgs)) {
+                pending = {};
+                pending_count = 0;
+                node.error(RED._("sort.too-many"));
+            }
+        }
+        
+        this.on("input", function(msg) {
+            process_msg(msg);
+        });
+    }
+
```

Comment:
do we need an on.close method to tidy up any outstanding parts cleanly ?

Reply:
Thank you for pointing out.
I think that is safer.